File: hack/verify-publishing-bot.py
#!/usr/bin/env python3

# Copyright 2022 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import os
import sys


def get_gomod_dependencies(rootdir: str, components: list[str]):
    all_dependencies = {}
    for component in components:
        with open(os.path.join(rootdir, component, "go.mod")) as f:
            print(component, "dependencies")
            all_dependencies[component] = []
            lines = sorted(set(f))
            for line in lines:
                for dep in components:
                    if dep == component:
                        continue
                    if f"k8s.io/{dep} =>" not in line:
                        continue
                    print(f"\t{dep}")
                    if dep not in all_dependencies[component]:
                        all_dependencies[component].append(dep)
    return all_dependencies


def get_rules_dependencies(rules_file: str):
    import yaml
    with open(rules_file) as f:
        content = yaml.safe_load(f)
    return content


def main():
    file_dir = os.path.dirname(__file__)
    root_dir = os.path.abspath(os.path.join(file_dir, os.pardir))
    components_dir = os.path.join(root_dir, "staging/src/k8s.io/")
    print(root_dir, components_dir)
    components = sorted(os.listdir(components_dir))
    rules_file = os.path.join(root_dir, "staging/publishing/rules.yaml")
    try:
        import yaml
    except ImportError:
        print("Please install missing pyyaml module and re-run", sys.argv[0])
        sys.exit(1)

    rules_dependencies = get_rules_dependencies(rules_file)
    gomod_dependencies = get_gomod_dependencies(components_dir, components)

    processed_repos = []
    for rule in rules_dependencies["rules"]:
        branch = rule["branches"][0]

        # If this no longer exists in master
        if rule["destination"] not in gomod_dependencies:
            # Make sure we don't include a rule to publish it from master
            for branch in rule["branches"]:
                if branch["name"] == "master":
                    raise Exception("Cannot find master branch for destination", rule["destination"])
            # And skip validation of publishing rules for it
            continue

        if branch["name"] != "master":
            raise Exception("Cannot find master branch for destination", rule["destination"])
        if branch["source"]["branch"] != "master":
            raise Exception("Cannot find master source branch for destination", rule["destination"])

        # we specify the go version for all master branches through `default-go-version`
        # so ensure we don"t specify explicit go version for master branch in rules
        if "go" in branch:
            raise Exception("go version must not be specified for master branch for destination", rule["destination"])

        print("Processing:", rule["destination"])
        if rule["destination"] not in gomod_dependencies:
            raise Exception("missing go.mod for", rule["destination"])

        processed_repos.append(rule["destination"])
        processed_deps = []
        for dep in set(gomod_dependencies[rule["destination"]]):
            found = False
            if "dependencies" in branch:
                for dep2 in branch["dependencies"]:
                    processed_deps.append(dep2["repository"])
                    if dep2["branch"] != "master":
                        raise Exception(
                            f"Looking for master branch and found: {dep2} for destination {rule['destination']}",
                        )
                    if dep2["repository"] == dep:
                        found = True
            else:
                raise Exception(
                    f"Please add {gomod_dependencies[rule['destination']]} as dependencies under destination {rule['destination']} in {rules_file}"
                )

            if found:
                print(f"Found dependency {dep}")
            else:
                raise Exception(
                    f"Please add {dep} as a dependency under destination {rule['destination']} in {rules_file}"
                )

        extraDeps = set(processed_deps) - set(gomod_dependencies[rule["destination"]])
        if len(extraDeps) > 0:
            raise Exception(
                f"Extra dependencies in rules for {rule['destination']}: {','.join(str(s) for s in extraDeps)}"
            )

    items = set(gomod_dependencies.keys()) - set(processed_repos)
    if len(items) > 0:
        raise Exception(f"Missing rules for {','.join(str(s) for s in items)}")
    print("Done")


if __name__ == "__main__":
    main()

File: hack/tools/verify/verify-publishing-bot.go
/*
Copyright 2023 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v2"
)

func getGomodDependencies(rootdir string, components []string) map[string][]string {
	allDependencies := make(map[string][]string)
	for _, component := range components {
		filePath := filepath.Join(rootdir, component, "go.mod")
		file, err := os.Open(filePath)
		if err != nil {
			log.Fatalf("Failed to open go.mod file for component %s: %v", component, err)
		}
		defer file.Close()

		fmt.Printf("%s dependencies\n", component)
		allDependencies[component] = []string{}

		content, err := ioutil.ReadAll(file)
		if err != nil {
			log.Fatalf("Failed to read go.mod file for component %s: %v", component, err)
		}

		lines := strings.Split(string(content), "\n")
		uniqueLines := make(map[string]bool)
		for _, line := range lines {
			uniqueLines[line] = true
		}

		for _, line := range lines {
			for _, dep := range components {
				if dep == component {
					continue
				}
				if !strings.Contains(line, "k8s.io/"+dep+" =>") {
					continue
				}
				fmt.Printf("\t%s\n", dep)
				if !contains(allDependencies[component], dep) {
					allDependencies[component] = append(allDependencies[component], dep)
				}
			}
		}
	}
	return allDependencies
}

func getRulesDependencies(rulesFile string) map[string]interface{} {
	data := make(map[string]interface{})
	file, err := os.Open(rulesFile)
	if err != nil {
		log.Fatalf("Failed to open rules file: %v", err)
	}
	defer file.Close()

	content, err := ioutil.ReadAll(file)
	if err != nil {
		log.Fatalf("Failed to read rules file: %v", err)
	}

	err = yaml.Unmarshal(content, &data)
	if err != nil {
		log.Fatalf("Failed to parse rules file: %v", err)
	}

	return data
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func main() {
	rootdir := filepath.Join(filepath.Dir(os.Args[0]), "/../../../")
	rootdir, err := filepath.Abs(rootdir)
	if err != nil {
		log.Fatalf("Failed to get absolute path: %v", err)
	}

	components := []string{}
	files, err := ioutil.ReadDir(filepath.Join(rootdir, "staging/src/k8s.io/"))
	if err != nil {
		log.Fatalf("Failed to read components directory: %v", err)
	}
	for _, file := range files {
		if file.IsDir() {
			components = append(components, file.Name())
		}
	}

	rulesFile := filepath.Join(rootdir, "staging/publishing/rules.yaml")
	rulesDependencies := getRulesDependencies(rulesFile)
	gomodDependencies := getGomodDependencies(filepath.Join(rootdir, "staging/src/k8s.io/"), components)

	processedRepos := []string{}
	for _, rule := range rulesDependencies["rules"].([]interface{}) {
		ruleData := rule.(map[interface{}]interface{})
		destination := ruleData["destination"].(string)
		branches := ruleData["branches"].([]interface{})
		branch := branches[0].(map[interface{}]interface{})

		if _, exists := gomodDependencies[destination]; !exists {
			// Make sure we don't include a rule to publish it from master
			for _, b := range branches {
				branchData := b.(map[interface{}]interface{})
				branchName := branchData["name"].(string)
				if branchName == "master" {
					log.Fatalf("Cannot find master branch for destination %s", destination)
				}
			}
			// And skip validation of publishing rules for it
			continue
		}

		for _, b := range branches {
			branchData := b.(map[interface{}]interface{})
			sourceDir := branchData["source"].(map[interface{}]interface{})["dir"].(string)

			fmt.Println()
			fmt.Println()
			fmt.Println()
			fmt.Println()
			fmt.Println(sourceDir)
			fmt.Println()
			fmt.Println()
			fmt.Println()
			fmt.Println()

			if sourceDir != "" {
				log.Fatalf("use of deprecated `dir` field in rules for `%s`", destination)
			}
			if len(branchData["source"].(map[interface{}]interface{})["dirs"].(string)) > 1 {
				log.Fatalf("cannot have more than one directory (`%s`) per source branch `%s` of `%s`",
					(branchData["source"].(map[interface{}]interface{})["dirs"].(string)),
					(branchData["source"].(map[interface{}]interface{})["branch"]),
					destination,
				)
			}

			if !strings.HasSuffix(sourceDir, destination) {
				log.Fatalf("Copy/paste error `%s` refers to `%s`", destination, sourceDir)
			}
		}

		branchName := branch["name"].(string)
		if branchName != "master" {
			log.Fatalf("Cannot find master branch for destination %s", destination)
		}

		sourceBranch := branch["source"].(map[interface{}]interface{})["branch"].(string)
		if sourceBranch != "master" {
			log.Fatalf("Cannot find master source branch for destination %s", destination)
		}

		if _, exists := branch["go"]; exists {
			log.Fatalf("Go version must not be specified for master branch for destination %s", destination)
		}

		fmt.Printf("processing : %s\n", destination)
		if _, exists := gomodDependencies[destination]; !exists {
			log.Fatalf("Missing go.mod for %s", destination)
		}
		processedRepos = append(processedRepos, destination)
		processedDeps := []string{}
		for _, dep := range gomodDependencies[destination] {
			found := false
			if _, exists := branch["dependencies"]; exists {
				dependencies := branch["dependencies"].([]interface{})
				for _, depData := range dependencies {
					depMap := depData.(map[interface{}]interface{})
					depRepo := depMap["repository"].(string)
					depBranch := depMap["branch"].(string)
					if depBranch != "master" {
						log.Fatalf("Looking for master branch and found: %s for destination %s", depBranch, destination)
					}
					if depRepo == dep {
						found = true
						break
					}
				}
			} else {
				log.Fatalf("Please add %s as dependencies under destination %s in %s", gomodDependencies[destination], destination, rulesFile)
			}

			if !found {
				log.Fatalf("Please add %s as a dependency under destination %s in %s", dep, destination, rulesFile)
			} else {
				fmt.Printf("  found dependency %s\n", dep)
			}
			processedDeps = append(processedDeps, dep)
		}

		extraDeps := []string{}
		for _, dep := range processedDeps {
			if !contains(gomodDependencies[destination], dep) {
				extraDeps = append(extraDeps, dep)
			}
		}

		if len(extraDeps) > 0 {
			log.Fatalf("Extra dependencies in rules for %s: %s", destination, strings.Join(extraDeps, ","))
		}
	}

	items := []string{}
	for k := range gomodDependencies {
		if !contains(processedRepos, k) {
			items = append(items, k)
		}
	}
	if len(items) > 0 {
		log.Fatalf("Missing rules for %s", strings.Join(items, ","))
	}

	fmt.Println("Done.")
}
